### PR TITLE
8262731: [macOS] Exception from "Printable.print" is swallowed during "PrinterJob.print"

### PR DESCRIPTION
**Pull request by alitvinv**
Hello,

Could you please review the following fix for the bug specific to macOS. The bug consists in the fact that if the method "java.awt.print.Printable.print​(Graphics, PageFormat, int)" throws "java.awt.print.PrinterException" or "java.lang.RuntimeException" during the call "java.awt.print.PrinterJob.print()", then the exception is caught and ignored by JDK and a user cannot learn that printing failed and what caused failure of printing, because "PrinterJob.print()" method does not throw "PrinterException" or the occurred exception is not reported by JDK through the error stream.

ROOT CAUSE OF THE BUG:
The root cause of the bug is the fact that in the method "sun.lwawt.macosx.CPrinterJob.printAndGetPageFormatArea(final Printable, final Graphics, final PageFormat, final int)" from the file "src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java" the exception thrown during execution of the expression

"int pageResult = printable.print(graphics, pageFormat, pageIndex);"

is caught but is not returned to a developer by any mean or is not printed out to the error stream.

THE FIX:
The fix implements propagation of the occurred and caught exception to the level of the user's code executing "PrinterJob.print()" method. Propagation of the exception by storing it in the instance variable of "CPrinterJob" object is implemented, because the engaged code always is executed:
- on 2 threads (non-EDT thread, EDT thread) in case when "PrinterJob.print()" is called by the user on a non-EDT thread;
- on 3 threads (2 EDT threads, a temporary thread started by JDK to execute "CPrinterJob._safePrintLoop(long, long );") when "PrinterJob.print()" is called on EDT thread.

The regression test which is part of the fix was also successfully executed on MS Windows OS and Linux OS.

Thank you,
Anton

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262731](https://bugs.openjdk.java.net/browse/JDK-8262731): [macOS] Exception from "Printable.print" is swallowed during "PrinterJob.print"

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4036/head:pull/4036` \
`$ git checkout pull/4036`

Update a local copy of the PR: \
`$ git checkout pull/4036` \
`$ git pull https://git.openjdk.java.net/jdk pull/4036/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4036`

View PR using the GUI difftool: \
`$ git pr show -t 4036`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4036.diff">https://git.openjdk.java.net/jdk/pull/4036.diff</a>

</details>

### CODE REVIEW

**Comment by alitvinv** [comment]
Publishing this RFR on the e-mail alias (2d-dev@openjdk.java.net) by writing this comment. Dear reviewers from (2d-dev@openjdk.java.net) e-mail alias, could you please review this RFR.

**Review by prrace — COMMENTED** [review]
I'm a bit surprised the test passes on windows.
Perhaps because you aren't trying RuntimeException ?

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            @Override
+            public int print(Graphics graphics, PageFormat pageFormat,
+                    int pageIndex) throws PrinterException {
+                if (pageIndex > 1) {
+                    return NO_SUCH_PAGE;
+                }
+                throw new PrinterException("Exception from Printable.print");
+            }
+        });
+        if (job.printDialog()) {
```
[the remark this answers is not in the document]
Comment (alitvinv) [reply test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:86]: Hello Phil. Thank you very much for review of this fix. First I am answering your question, which is about why the test passes on Windows and which is your first and separate comment not bound to the source code. The test passes on Windows, because in JDK 17 for both Windows and Linux JDK works similarly stably and reliably, JDK 17 on Window and Linux successfully propagates "PrinterException" and "RuntimeException", if it is thrown from "Printable.print" method. I had verified this using my standalone manual test case "ExceptionFromPrintableIsIgnored.java" attached to the bug record in JBS, before sending the 1st version of the fix for review. I did not explore, why it is so, because it works on Windows and Linux and this bug is related to macOS, but I think that it works on Windows and Linux, because in JDK implementations for Windows and Linux printing is consistently done by JDK on the same thread on which "PrinterJob.print" was called and no portions of the involved code are executed asynchronously on other thread, like EDT in case of macOS.

I agree there is no need in the dialog in the regression test. In the 1st version of the fix I did it deliberately, to make it more manual, because the test initiates printing and there is no guarantee what type of printer would be set up on the test host, if some virtual printer is used on the host, that printer can show its own native dialog asking to specify location of a file, in which the printed document should be saved. That is why I think that this test cannot be fully automatic. In the 2nd version of the fix which I am submitting right now I removed this unnecessary code showing JDK print dialog.

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    private void runTest() {
+        PrinterJob job = PrinterJob.getPrinterJob();
+        job.setPrintable(new Printable() {
+            @Override
+            public int print(Graphics graphics, PageFormat pageFormat,
+                    int pageIndex) throws PrinterException {
+                if (pageIndex > 1) {
+                    return NO_SUCH_PAGE;
+                }
+                throw new PrinterException("Exception from Printable.print");
```
[the remark this answers is not in the document]
Comment (alitvinv) [reply test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:83]: In the 1st version of the fix I did not test the scenario with "RuntimeException", because despite of the fact that I had known that JDK 17 will pass on Windows and Linux the scenario with "RuntimeException" I still was not sure if the behavior of older JDK release families is the same with JDK 17 on Windows and Linux, and just did not want to run the risk of failure of this test during potential porting of the fix to older JDK release families.

But yesterday I did testing with my standalone test case attached to the bug record in JBS and now I can certainly say that JDK 8u291-b10, JDK JDK 8-b132 do not fail in the test scenarios with "RuntimeException" on Windows and Linux. Therefore it is safe to include test scenario with "RuntimeException" to the regression test.

In the 2nd version of the fix now I am testing also "RuntimeException" both on EDT thread and on main thread.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java [hunk 3]
```diff
@@ -256,6 +258,14 @@ private void completePrintLoop() {
         }
     }
 
+    private Exception setLastPrintEx(Exception newEx, boolean printOldEx) {
+        Exception oldEx = lastPrintExRef.getAndSet(newEx);
+        if (printOldEx && (oldEx != null)) {
+            oldEx.printStackTrace();
```
[the remark this answers is not in the document]
Comment (alitvinv) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:264]: Thank you for this remark. I agree using "Throwable" is more solid approach. In the 2nd version of the fix I use "Throwable" in this "AtomicReference" object and I removed completely this method "setLastPrintEx".

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java [hunk 4]
```diff
@@ -785,22 +809,37 @@ private void printToPathGraphics(    final PeekGraphics graphics, // Always an a
...
+            public void run() {
+                synchronized (ret) {
+                    try {
+                        int pageResult = printable.print(
+                            graphics, pageFormat, pageIndex);
+                        if (pageResult != Printable.NO_SUCH_PAGE) {
+                            ret[0] = getPageFormatArea(pageFormat);
+                        }
+                    } catch (Exception e) {
+                        // Original code bailed on any exception
```
[the remark this answers is not in the document]
Comment (alitvinv) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:823]: In the 2nd version of the fix I substituted "Exception" for "Throwable" in this catch block and in second "catch" block several lines below. Also in the 2nd fix version I removed this comment about bailing.

My opinion is the same, the code before the fix did not work according to the specification and was ignoring any caught exceptions. Perhaps, the reason of this is involvement of more than 1 thread in the code flow standing behind "PrinterJob.print" method execution and execution of the JDK's code through JNI in this code flow. Thus propagating the exception is impossible across threads and it is not easy to stop printing done by macOS at the moment, when the exception occurs in Java code in JDK. This pattern is used in the file "CPrinterJob.java" in other places, even there is the same comment in the method "CPrinterJob.getPageformatPrintablePeekgraphics". At the same time propagating "RuntimeException" from "PrinterJob.print" on Windows, Linux and now on macOS after this fix also is not according to the specification but still consistency in JDK behavior on these OS versions should be good.

As I understand, the method "CPrinterJob.printAndGetPageFormatArea" returns "null" as the "Rectangle2D", when the exception is caught, and this "null" value should automatically lead to returning of "NSZeroRect" from the Objective-C method "(NSRect)rectForPage:(NSInteger)pageNumber" in the file "src/java.desktop/macosx/native/libawt_lwawt/awt/PrinterView.m", from which "CPrinterJob.printAndGetPageFormatArea" Java method was invoked. And this "NSZeroRect" return value is indication of the error for macOS, which further leads to stopping of the printing process.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java [hunk 5]
```diff
@@ -380,6 +391,19 @@ public SecondaryLoop run() {
             if (printingLoop != null) {
                 printingLoop.exit();
             }
+
+            Exception lastPrintEx = setLastPrintEx(null, false);
+            if (lastPrintEx != null) {
+                if (lastPrintEx instanceof PrinterException) {
+                    throw (PrinterException) lastPrintEx;
+                } else if (lastPrintEx instanceof RuntimeException) {
+                    throw (RuntimeException) lastPrintEx;
```
[the remark this answers is not in the document]
Comment (alitvinv) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:400]: In the 2nd version of the fix I added testing of the scenario with "RuntimeException".

Yes, this code is executed only after the execution of the "printLoop" is fully finished for each of all page ranges. If printing of one page fails within one page range, then printing of other pages does not occur. "CPrinterJob.print(PrintRequestAttributeSet)" is blocking until printing is finished or fails, and my code is executed in the very end of the method.

**Comment by alitvinv** [comment]
Hello Phil,

I have uploaded the 2nd version of the fix, which addresses your remarks. Could you please review it, also I provided answers to all your comments. Changes in the 2nd version of the fix are following:

1. The created atomic reference object now stores "Throwable" instance instead of "Exception", the variable itself is renamed to "printErrorRef".
2. The method "setLastPrintEx" introduced by the 1st version of the fix is completely removed.
3. "catch" blocks in the method "CPrinterJob.printAndGetPageFormatArea" now catches "Throwable" instead of "Exception".
4. The regression test now tests also "RuntimeException" on EDT and on the main thread.
5. The regression test now does not show any JDK print dialog.

Thank you,
Anton

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+/* @test
+   @bug 8262731
+   @key headful printer
+   @summary Verify that "PrinterJob.print" throws the expected exception,
+            if "Printable.print" throws an exception.
+   @run main/manual ExceptionFromPrintableIsIgnoredTest MAIN PE
+   @run main/manual ExceptionFromPrintableIsIgnoredTest MAIN RE
+   @run main/manual ExceptionFromPrintableIsIgnoredTest EDT PE
+   @run main/manual ExceptionFromPrintableIsIgnoredTest EDT RE
```
Comment (prrace) [inline test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:32]: wjy is this still manual ?

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                throw new RuntimeException(e);
+            }
+        }
+        System.out.println("Test passed.");
+    }
+
+    private void runTest(final TestExceptionType exceptionType) {
+        PrinterJob job = PrinterJob.getPrinterJob();
+        if (job.getPrintService() == null) {
+            throw new RuntimeException("No printers are available.");
```
Comment (prrace) [inline test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:85]: where does this exception go if it happens and you are on EDT ?

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            printEx.printStackTrace(System.out);
+
+            if (((printEx instanceof PrinterException) &&
+                    (exceptionType == TestExceptionType.PE)) ||
+                ((printEx instanceof RuntimeException) &&
+                    (exceptionType == TestExceptionType.RE))) {
+                return;// Test passed.
+            }
+            throw new RuntimeException("Unexpected exception was thrown.");
+        } else {
```
Comment (prrace) [inline test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:124]: When running on the EDT do you really want to do this ?

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        if (printEx != null) {
+            System.out.println("'PrinterJob.print' threw the exception:");
+            printEx.printStackTrace(System.out);
+
+            if (((printEx instanceof PrinterException) &&
+                    (exceptionType == TestExceptionType.PE)) ||
+                ((printEx instanceof RuntimeException) &&
+                    (exceptionType == TestExceptionType.RE))) {
```
Comment (prrace) [inline test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:120]: 1) I am fairly sure that I tested this case on Windows and nothing caught it so does this test work on Windows ?

2) I raised the question / suggestion that should we not wrap these in a PrinterException ? 
It at least needs to be thought through.

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java (hunk 6 above)
Comment (alitvinv) [reply test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:32]: The test is manual, because the test initiates printing and there is a chance that on a test host a default printer will be some virtual printer which can show the native dialog asking to specify the location of PDF file in which the printed document should be saved, in this case the test will be blocked and will be killed by "jtreg" by a timeout, what is unacceptable for the automatic test which should not slow down the speed of execution of all other automatic tests. For example such a virtual printer on Windows OS can be "Microsoft Print to PDF". By the way, I already explained this reason in one of my replies to your questions in the previous iteration of the code review.

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java (hunk 7 above)
Comment (alitvinv) [reply test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:85]: If this exception happens on EDT, this exception as any other exception thrown from the method "runTest" will be perfectly caught and handled in "catch" block of the constructor of the test "ExceptionFromPrintableIsIgnoredTest" as "InvocationTargetException" instance, which will have this "RuntimeException" as its cause. The whole test relies on this approach of throwing and catching the exception and it works.

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java (hunk 8 above)
Comment (alitvinv) [reply test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:124]: Yes, I do really want to do this, when running on EDT, the whole test relies on this approach. When the test is run on EDT this is done via "SwingUtilities.invokeAndWait" method in the constructor of the test "ExceptionFromPrintableIsIgnoredTest" and in that constructor there is "try/catch" block. If some exception occurs on EDT and is not caught on EDT, then it will be wrapped by JDK in "InvocationTargetException" instance which will be thrown from "SwingUtilities.invokeAndWait" method and will be caught by that "try/catch" block in the constructor of the test. I checked that this code works before sending the code for review and today additionally.

File: test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java (hunk 9 above)
Comment (alitvinv) [reply test/jdk/java/awt/print/PrinterJob/ExceptionFromPrintableIsIgnoredTest.java:120]: 1. The test does not fail on Windows in all 4 tested scenarios. You can run this regression test on Windows and you will see in the test report in the test output that the test successfully throws the required "PrinterException" or "RuntimeException" from "java.awt.print.Printable.print​(Graphics, PageFormat, int)" and that the test successfully catches this "PrinterException" or "RuntimeException" and prints out it to "System.out" "PrintStream" in the test method "ExceptionFromPrintableIsIgnoredTest.runTest(final TestExceptionType)".

Line 115    printEx.printStackTrace(System.out);

This line in the test is done exactly as the evidence that the exception was really propagated from the method "PrinterJob.print()" and caught by the test. I verified that it is "true" on macOS, Windows, Linux.

I reported this in the code review and in JBS bug record itself that in JDK 17 and in JDK 8u291 and even in JDK 8-b132 on Windows and on Linux "PrinterJob.print()", which is called on EDT, propagates "PrinterException" and "RuntimeException" which are thrown from "Printable.print​(Graphics, PageFormat, int)" method. If you still do not trust this, then I suggest you just to recheck it yourself by any trustworthy mean in addition to this regression test and the purely manual test case "ExceptionFromPrintableIsIgnored.java" attached to the bug record.

2. By current moment in JDK versions on Windows and Linux the "RuntimeException" thrown from "Printable.print​(Graphics, PageFormat, int)" has been propagated through "PrinterJob.print()" as is without wrapping it into "PrinterException", as my experiments show even in JDK 8-b132 and in currently latest GA release of JDK 8 and in JDK 17.
 
From my viewpoint not wrapping "RuntimeException" into "PrinterException" and propagating "RuntimeException" through "PrinterJob.print()" will just bring behavior of JDK on macOS into correspondence with existing behavior of JDK on Windows and Linux. If we wrap "RuntimeException" into "PrinterException" on macOS, then at some point in time somebody will need to change JDK code for Windows and Linux to also do the same wrapping of the exception which is currently not done. I do not insist on any variant.

**Review by prrace — COMMENTED** [review]
The points I have been trying to make are that
1) We need to make sure that important cleanup happens rather than just bailing
2) Despite Windows + Linux propagating a RTE, I am not sure this is what we want.
   If the RTE is something the *app* decides to throw,  then that's one thing but
   if something went wrong in the bowels of printing the spec says it should be a PrinterException.
   At the very least we should make sure temp files are deleted, any open calls
   to startDoc have been matched with a close ..

Hence my early comment "The exception should not propagate. State may need to be cleaned up. 
If you do anything you would make it the init cause of the PrinterException."
That means if we see an RTE propagated .. that could be a bug.

So I am nor comfortable just "aligning" the behaviour with WIndows and Linux without being
sure they are correct.

In other words, mac was probably catching this to do the clean up part, and that was likely OK
But yes, it probably was wrong not to rethrow as a wrapped (if needed) PrinterException

But although it opens a whole new can of worms I was suggesting you needed to look closely
at windows + linux and not just create a test which verified behaviour I am not sure we want.

I questioned the test etc about RTE because although it is good to test hrowing RTE  you
are verifying it ARRIVES still as an RTE, and I think it should be a PrinterException.

I am open to discussion about why this may be "actually OK" but it has to be better than
"the others are doing it" and it definitely should not be a reason to leave open files/jobs.

**Comment by alitvinv** [comment]
Hello Phil,

I created the 3rd version of the fix which should address your remarks about the fix and the regression test. Can you please review the 3rd version of the fix (Webrev 03).

Thank you for provision with your position about different aspects of the fix, regression test and the discovered discrepancies with the specification on Windows OS and Linux OS. I absolutely agree that it is incorrect to implement propagation of RTE through "PrinterJob.print()" without wrapping it into "PrinterException", because RTE is not mentioned in the specification for "PrinterJob.print()" method. The 3rd fix version addresses this mistake from previous fix versions.

Perhaps, it is not very significant issue, that on Windows OS and Linux OS RTE from "Printable.print​(Graphics, PageFormat, int)" is propagated, because in this case the user's code does not follow the specification by throwing RTE from "Printable.print​(Graphics, PageFormat, int)". I do not want to investigate this discovered discrepancy with specification on Windows OS and Linux OS in this fix for a little different issue specific to macOS. If it is currently really necessary I can file a bug or 2 bugs in JBS to address the discovered issue on Windows and Linux.

Thank you for giving me the details about "printer" value for "@key" tag in "jtreg" test, because until that moment I mistakenly thought that it was just a marker used for selection of tests marked with it for running and nothing else.

DIFFERENCES INTRODUCED IN THE 3'RD VERSION OF THE FIX:

THE FIX:
1. "RuntimeException" is wrapped in new "PrinterException" in "CPrinterJob.print(PrintRequestAttributeSet)" to guarantee that the fix can propagate only "PrinterException" from "PrinterJob.print()".

THE REGRESSION TEST:
1. The test is now fully automatic, "manual" is removed from all JTREG "main" actions. I verified it on MACH and locally.

2. I addressed your concerns from previous comments and now no exception is thrown from "ExceptionFromPrintableIsIgnoredTest.runTest" method, so there is no any possibility that AWT thread will be crashed by the exception thrown by the test. Now I just execute "PrinterJob.print()" and store any "Throwable" possibly thrown by that method in the instance variable "private volatile Throwable printError" of the test class. And analyze it always on main thread after completion of execution of "ExceptionFromPrintableIsIgnoredTest.runTest" call.

3. Now the regression test always expects only "PrinterException" to be thrown by "PrinterJob.print()". Since "RuntimeException" is propagated from "PrinterJob.print()" on Windows OS and on Linux OS two test scenarios with "RuntimeException":
- "@run main ExceptionFromPrintableIsIgnoredTest MAIN RE"
- "@run main ExceptionFromPrintableIsIgnoredTest EDT RE"

are not executed by the test on Windows and Linux deliberately, otherwise they are failing. This test restriction is implemented by the "if" condition "if ((exceptionType == TestExceptionType.RE) && !isOSX) {".

Hopefully this version of the fix is more acceptable.

Thank you,
Anton

**Review by prrace — APPROVED** [review]
Ok. Please file a bug as you mention to bring the other platforms into line.

**Comment by alitvinv** [comment]
Hello Phil. Thank you for approval of the fix. The new bug about the discovered propagation of the RTE on Windows OS and Linux OS is created, it is (https://bugs.openjdk.java.net/browse/JDK-8268675).

Thank you,
Anton

________________________________________
From: prrace ***@***.***>
Sent: 11 June 2021 20:47
To: openjdk/jdk
Cc: Anton Litvinov; Mention
Subject: [External] : Re: [openjdk/jdk] 8262731: [macOS] Exception from "Printable.print" is swallowed during "PrinterJob.print" (#4036)

@prrace approved this pull request.

Ok. Please file a bug as you mention to bring the other platforms into line.
